server/resp/src/main/java/org/infinispan/server/resp/RespLettuceHandler.java

```diff
@@ -19,18 +22,53 @@ public class RespLettuceHandler extends ByteToMessageDecoder {
 
    private final RedisStateMachine stateMachine = new RedisStateMachine(ByteBufAllocator.DEFAULT);
    private RespRequestHandler requestHandler;
+   private boolean disabledRead = false;
 
    public RespLettuceHandler(RespRequestHandler initialHandler) {
       this.requestHandler = initialHandler;
    }
 
    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
+      // Don't read any of the ByteBuf if we disabled reads
+      if (disabledRead) {
+         return;
+      }
       PushOutput<byte[], byte[]> pushOutput = new PushOutput<>(ByteArrayCodec.INSTANCE);
       if (stateMachine.decode(in, pushOutput)) {
          String type = pushOutput.getType().toUpperCase();
          List content = pushOutput.getContent();
-         requestHandler = requestHandler.handleRequest(ctx, type, content.subList(1, content.size()));
+         List<byte[]> contentToUse = content.subList(1, content.size());
+         log.tracef("Received command: %s with arguments %s", type, Util.toStr(contentToUse));
+         CompletionStage<RespRequestHandler> stage = requestHandler.handleRequest(ctx, type, contentToUse);
+         if (CompletionStages.isCompletedSuccessfully(stage)) {
+            requestHandler = CompletionStages.join(stage);
+         } else {
+            log.tracef("Disabling auto read for channel %s until previous command is complete", ctx.channel());
+            // Disable reading any more from socket - until command is complete
+            ctx.channel().config().setAutoRead(false);
+            disabledRead = true;
+            stage.whenComplete((handler, t) -> {
+               assert ctx.channel().eventLoop().inEventLoop();
+               log.tracef("Re-enabling auto read for channel %s as previous command is complete", ctx.channel());
+               ctx.channel().config().setAutoRead(true);
+               disabledRead = false;
+               if (t != null) {
+                  exceptionCaught(ctx, t);
+                  return;
+               }
```

Comment:
By returning here we could end up losing some bytes?

Reply:
Ah, true. We should check the internal buffer still 👍